### PR TITLE
Update due to keystone external pkg move

### PR DESCRIPTION
**Pull request by stuggi**
update due to [1] which moves the content of the external pkg to the api sub module to remove the dependency on the keystone-operator internals and just have to import the api.

Also updates the import of lib-common to the current latest.

For the envtest to get the required 3rd party operators CRDs, the local go.mod is checked for the operator api module version, creates a tmp go.mod file to load the corresponding 3rd party main operator module.

[1] https://github.com/openstack-k8s-operators/keystone-operator/pull/90

### CODE REVIEW

File: controllers/placementapi_controller.go [hunk 1]
```diff
@@ -31,7 +31,6 @@ import (
 	"github.com/go-logr/logr"
 	routev1 "github.com/openshift/api/route/v1"
 	keystonev1 "github.com/openstack-k8s-operators/keystone-operator/api/v1beta1"
```
Comment (SeanMooney) [inline controllers/placementapi_controller.go:33]: nit: you could have kept this as just keystone

File: controllers/suite_test.go [hunk 2]
```diff
@@ -67,30 +71,88 @@ func TestAPIs(t *testing.T) {
 		[]Reporter{printer.NewlineReporter{}})
 }
 
-func GetDependencyVersion(moduleName string) (string, error) {
+func GetDependencyVersion(moduleName string) (string, string, error) {
```
Comment (SeanMooney) [inline controllers/suite_test.go:74]: its too bad these return type are not named
in isolation its impossible to know what the first and second return argument
are without treading the body fo the function 
 name, version, nil

Comment (SeanMooney) [reply controllers/suite_test.go:74]: looking at https://www.geeksforgeeks.org/named-return-parameters-in-golang/
we are allowed to name them so please do.

i.e.

func add(x int, y int) (int) {
	return x + y
}


func add(x int, y int) (ret int) {
	return x + y
}

func add(x int, y int) (ret int) {
    ret = x+y
	return
}

are all legal and do the same thing but the second form is the most readble

so we should use 
func GetDependencyVersion(moduleName string) (name string, version string, err error) {

File: controllers/placementapi_controller.go (hunk 1 above)
Comment (stuggi) [reply controllers/placementapi_controller.go:33]: its common when using APIs to name it using the major version.

File: controllers/suite_test.go (hunk 2 above)
Comment (stuggi) [reply controllers/suite_test.go:74]: ack will do

Comment (stuggi) [reply controllers/suite_test.go]: done

File: controllers/placementapi_controller.go (hunk 1 above)
Comment (SeanMooney) [reply controllers/placementapi_controller.go:33]: ack its fine either way i just called it out as you could have avoided renameing it below but if we are only gong to rev that when we rev a major version and not everytime we change a minor version I'm totally fine with that

File: controllers/suite_test.go (hunk 2 above)
Comment (SeanMooney) [reply controllers/suite_test.go]: thanks

**Review by SeanMooney — APPROVED** [review]
this looks good to me but lets see what dan and gibi think

**Comment by gibizer** [comment]
I have concerns:
- 1) I'm not sure that a version number from keystone/api will always exists for keystone. Imagine if we only fix something in the API definition (i.e. api documentation) and release that as v1.0.1 from keystone/api but as it does not required any implementation change the latest keystone module version is kept as v1.0.0. So this approach will only work if we always release keystone/api and keystone together.
- 2) This is now a lot of hoops to jump to get something that we could be easily get by storing the API Structs and the generated CRD at the same module.
- 3) Following from 2) If we store the structs that defines the CRD in keystone/api then I think we should also store the CRD that is  directly generated from those structs in the same module. As the API structs also the CRD is part of the API definition.

**Review by SeanMooney — DISMISSED** [review]
i dont know how to change approvals but this only had the first 3 commits when i reviewed it last
---- latere
no that's not right instead fo rebasing and fixing my comment in the commit martin added the 4th commit and i just didn't see that in the UI because github sucks for review.

**Comment by SeanMooney** [comment]
https://stackoverflow.com/a/72813625

i dont like that if you change the pr after you review the approval flag is not cleard

**Comment by SeanMooney** [comment]
> I have concerns:
> 
>     * 1. I'm not sure that a version number from keystone/api will always exists for keystone. Imagine if we only fix something in the API definition (i.e. api documentation) and release that as v1.0.1 from keystone/api but as it does not required any implementation change the latest keystone module version is kept as v1.0.0. So this approach will only work if we always release keystone/api and keystone together.
> 
>     * 2. This is now a lot of hoops to jump to get something that we could be easily get by storing the API Structs and the generated CRD at the same module.
> 
>     * 3. Following from 2) If we store the structs that defines the CRD in keystone/api then I think we should also store the CRD that is  directly generated from those structs in the same module. As the API structs also the CRD is part of the API definition.

so i obviously misunderstood the first commit
i thought we were going to go with the symlinking approach and this was going to use the new version of the keystone operator where that had already been done?

and that is why we were getting the CRD form the API module

**Comment by stuggi** [comment]
> > I have concerns:
> > ```
> > * 1. I'm not sure that a version number from keystone/api will always exists for keystone. Imagine if we only fix something in the API definition (i.e. api documentation) and release that as v1.0.1 from keystone/api but as it does not required any implementation change the latest keystone module version is kept as v1.0.0. So this approach will only work if we always release keystone/api and keystone together.

This is a valid concern and yes requires both versions stay in sync

> > 
> > * 2. This is now a lot of hoops to jump to get something that we could be easily get by storing the API Structs and the generated CRD at the same module.
> > 
> > * 3. Following from 2) If we store the structs that defines the CRD in keystone/api then I think we should also store the CRD that is  directly generated from those structs in the same module. As the API structs also the CRD is part of the API definition.
> > ```
> 
> so i obviously misunderstood the first commit i thought we were going to go with the symlinking approach and this was going to use the new version of the keystone operator where that had already been done?

No, this PR was to demonstrate an optional way to instead of using the linked CRDs to load the base module with the information of the api mod from go.mod. Sorry if that was not clear.

**Comment by stuggi** [comment]
storing in the CRDs in /api and add a symlink from the original place to be able to use the default create mechanism is the easiest and cleanest way.

**Comment by stuggi** [comment]
I'll get a new one filed now that we have https://github.com/openstack-k8s-operators/keystone-operator/pull/102 and https://github.com/openstack-k8s-operators/mariadb-operator/pull/35

**Pull request closed without merging**